unified autobuild github-action

nefarius2001 opened the pull request:
Work for #854:
single workflow 
- compiling on all platforms, including codeQL where possible => nice check for each PR
- separate codeQL-builds where necessary
- upload packs/binaries/installers as artifacts of the job (autodelete after 31 days) => low barrier for testing features on different platforms
- upload packs/binaries/installers to releases if triggered by tags starting with "r_" << still needs small tweaking
- updating packs/binaries/installers in a special release if triggered by tag named "latest" << still needs small tweaking

ann0see commented:
Thank you very much for your work again ;-). Can you also include #964?

nefarius2001 commented:
Sure. See the merge here: https://github.com/nefarius2001/jamulus/tree/autobuild_head

nefarius2001 commented:
will need #962

nefarius2001 commented:
Finally, my work is done!

One open question: currently a pull request triggers code-QL warnings ("Please make sure that every branch in on.pull_request is also in on.push so that Code Scanning can compare pull requests against the state of the base branch")
Personally, I'd add "push.master" as a trigger as well, but in the discussion #854 being resourceful was highly empasised, too.

For proof of working, see:
https://github.com/nefarius2001/jamulus/releases
https://github.com/nefarius2001/jamulus/actions

checks will likely need retriggering, when the submodule issue is solved

pljones commented:
Hm.  Might need another rebase to pick up the submodule changes, I'm afraid.

nefarius2001 commented:
Repeating the action would suffice, can someone click „rerun all jobs“ please?

pljones commented:
I did.  It failed the same.

hoffie commented:
I think the rebase on master is required in order to have #985 included in this PR.

ann0see commented:
The url of the sub modules is wrong: https://github.com/**github.com**/softins/jamulus-web/

Not sure if this helps?

nefarius2001 commented:
I used a normal merge... less effort since i merged ann0see's contributions earlier

ann0see commented on .github/workflows/autobuild.yml:
```diff
@@ -6,20 +6,28 @@ on:
       - "latest"
-    #branches:
+    branches:     
+      - "autobuild*"
```
Not sure about this, but is this a debugging branch?

nefarius2001 replied:
yes. developers can use branches with that pattern, to follow closely if any commit breaks their build

adding a comment to clarify

ann0see replied:
Ok. Thanks.

ann0see commented on .github/workflows/autobuild.yml:
```diff
@@ -65,92 +73,190 @@ jobs:
+           - config_name:           AndroidAPK (artifact+codeql)
+             target_os:             android
+             building_on_os:        ubuntu-20.04
+             cmd1_prebuild:         "sh android/autobuild/autobuild_apk_1_prepare.sh "
```
Interesting and probably much cleaner approach ;-).

ann0see commented on .github/workflows/autobuild.yml:
```diff
@@ -65,92 +73,190 @@ jobs:
-             asset_name:            jamulus_${{ needs.create_release.outputs.version_name }}_android.apk
+             cmd1_prebuild:         powershell ./windows/autobuild/autobuild_windowsinstaller_1_prepare.ps1
+             cmd2_build:            powershell ./windows/autobuild/autobuild_windowsinstaller_2_build.ps1
+             cmd3_postbuild:        powershell ./windows/autobuild/autobuild_windowsinstaller_3_copy_files.ps1
```
I assume my PR (with the nsProcess building) is included

nefarius2001 replied:
yes, merged as 81ebb786c34c70d0c5ac78e39bcfca48acb4ed4a

ann0see replied:
Yes. Saw that too.

ann0see commented on .github/workflows/autobuild.yml:
```diff
@@ -65,92 +73,190 @@ jobs:
+        if:                         ${{ matrix.config.cmd1_prebuild }}
+        run:                        ${{ matrix.config.cmd1_prebuild }}
+        env:
+          ACTIONS_ALLOW_UNSECURE_COMMANDS: 'true'  # allow setting environment variables
```
Sounds a bit dangerous. Can't you just output environment variables via echo?

nefarius2001 replied:
this allos the use of ::set-env name=xxxxxx::${xxxxxxx}

My problem was that autobuild_apk_1_prepare.sh sets a lot of variables, and I dont know which ones are necessary for building then. Doint special treatment for 15 variables via output and mentioning them as environment of next step is not clean any more. I'm open for suggestions.

ann0see replied:
Hmm. If you think this is the best way to do it, I would just leave it like that for now.

ann0see commented on .github/workflows/autobuild.yml:
```diff
@@ -159,6 +265,6 @@ jobs:
-        asset_path:               /github/home/io.jamulus.Jamulus/io.jamulus.Jamulus.flatpak
-        asset_name:               io.jamulus.Jamulus.flatpak
+        asset_path:               ${{steps.flatpak-download.outputs.download-path}}/${{ steps.flatpak-prepare.outputs.flatpak_bundle }}
+        asset_name:               ${{ steps.flatpak-prepare.outputs.flatpak_name }}
```
I'm not sure, but what happens on a PR? I saw that the release job ran. I hope, it doesn't automatically do a release on every PR. 

Another thought: would it be possible that this also moves the latest tag automatically if a tag starting with "r" is pushed? Unfortunately, I think we can't really move away from this.

nefarius2001 replied:
1) building artifacts is triggered on every pull request. Code QL is done. Artifacts are published as a part 
2) releases are triggered only if ((is_tag)  && (pattern "r_*" or "latest")) => not by requests, not by branches
Where is a good place to make the two above points clear? Where would you look for documentation of that? as a first suggestion, I will prepend the autobuild.yml with a few lines

3) if you plan to pop out releases every second day, I will implement automatisation for that happily. for releases every week or month, this gives you the opportunity to verify the builds as r_x_x_x. Moving "latest" needs you to move the tag only. one line of git command. I'd suggest that - let me know if you are really eager to have that step (that one git tag move&push) automated.

ann0see replied:
1. Thanks for the clarification
2. I think that's enough
3. Unfortunately it's much more than one git command. You'll need to remove the latest tag locally, afterwards push the removed tag on GitHub, then re-add it locally and push it to github again. So it's much more than that unfortunately. I'd like to have this automated personally.

nefarius2001 replied:
@ann0see Is a release script already implemented? My suggestion would be to add those two commands to it:
git tag -f latest
git push --force github_nefarius latest;

ann0see replied:
No. A release script is not yet implemented. I think you could use my proposed shell script in another issue.

ann0see commented on android/autobuild/autobuild_apk_2_build.sh:
```diff
@@ -0,0 +1,34 @@
+    exit 1
+fi
+
+cd ${THIS_JAMULUS_PROJECT_PATH}
```
Could you please quote all bash variables? Codacy (at least on my repo) complains if there are no quotes. Also paths with spaces might work better then. Especially in the lines below.

pljones replied:
Applies to _all_ the bash scripts.

ann0see commented on android/autobuild/autobuild_apk_3_copy_files.sh:
```diff
@@ -0,0 +1,44 @@
+echo ""
+echo ""
+echo "ls GITROOT/android-build/build/outputs/apk/debug/"
+ls ${THIS_JAMULUS_PROJECT_PATH}/android-build/build/outputs/apk/debug/
```
Debug output?

ann0see commented on mac/autobuild/artifacts/autobuild_mac_2_build.sh:
```diff
@@ -0,0 +1,20 @@
+    exit 1
+fi
+
+cd ${THIS_JAMULUS_PROJECT_PATH}
```
I assume I wrote some of this a few months ago and I didn't know about the quoting issue. Could you please add quotes here?

pljones replied:
All the bash scripts need checking.

ann0see commented on windows/autobuild/autobuild_windowsinstaller_1_prepare.ps1:
```diff
@@ -1,7 +1,22 @@
 
-# Get the source path via parameter
-param ([string] $sourcepath)
+
```
Why do you need so many empty lines here?

nefarius2001 replied:
legacy. At one point, all scipts had parameters in line 1-20 and instructions in line 21-x
Adding Parameters/Procedure headers instead

nefarius2001 replied:
changed.

ann0see commented on windows/autobuild/autobuild_windowsinstaller_2_build.ps1:
```diff
@@ -0,0 +1,21 @@
+# Sanity check of parameters
+if(($jamulus_project_path -eq $null) -or ($jamulus_project_path -eq "")){
+	throw "expecting ""jamulus_project_path"" as parameter or ENV"
+}elseif (!(Test-Path -Path $jamulus_project_path)){
```
could you please add spaces here and stick to the code style of the old .ps1 files?

nefarius2001 replied:
What you are saying is basically "use 4 spaces for indent instead of 1 tab", correct? => can do, probably for all ps & sh-files?
Is there anything else in terms of "code-style" I'm not seeing?

ann0see replied:
Personally, I'd suggest two spaces, but corrados used four and I think we should stick to that. Also in the deploy_windows.ps1 file the style is different:

e.g.: 
```powershell
    if (Test-Path -Path $BuildPath) { Remove-Item -Path $BuildPath -Recurse -Force }
```

So you'd need to add a few more spaces here. 

Yes, in a few spaces the code style could be changed. E.g. all variables in the bash scripts, especially those who represent a path should be quoted: ${VAR} should be "${VAR}".

nefarius2001 replied:
Its good to stick to established style. (My personal preference is tabs, then everyone can set their editor to show them what they like. and one indent=one character.)
I did what I saw to be consistent. If theres something I'm missing, tell me.

Does the quoting of path variable concern Powershell only, of bash, too?

ann0see replied:
Yes. Bash is important too.

ann0see commented on windows/autobuild/autobuild_windowsinstaller_2_build.ps1:
```diff
@@ -0,0 +1,21 @@
+)
+# Sanity check of parameters
+if(($jamulus_project_path -eq $null) -or ($jamulus_project_path -eq "")){
+	throw "expecting ""jamulus_project_path"" as parameter or ENV"
```
Not sure about the double quotes here? Don't you need + or . to add the content of the jamulus_project_path variable here? Not a PS expert though.

nefarius2001 replied:
this errormessage shall contain the variable-name nof whats missing, so no . / + here

ann0see replied:
Ok. So are two `"` the right syntax here?

nefarius2001 replied:
yes.
results in this message:
![grafik](https://user-images.githubusercontent.com/6545169/107886169-84946b80-6efe-11eb-9dbd-5f0344169b09.png)

ann0see commented on windows/autobuild/autobuild_windowsinstaller_3_copy_files.ps1:
```diff
@@ -0,0 +1,48 @@
+}else{
+	echo "jamulus_project_path is valid: $jamulus_project_path"
+}
+if(($jamulus_buildversionstring -eq $null) -or ($jamulus_buildversionstring -eq "")){
```
Spacing as above.

nefarius2001 replied:
resolved?

ann0see replied:
Probably. Will do a review soon.

ann0see commented on windows/deploy_windows.ps1:
```diff
@@ -275,7 +271,32 @@ Function Build-Installer
 }
 
+# Build and copy NS-Process dll
+Function Build-NSProcess
```
Ok. I now see that you've implemented my PR ;-). Thanks!

nefarius2001 replied:
so now I see you've found the implementation already :)

ann0see reviewed (changes requested):
First a few overview. I'll look into the functionality soon.

ann0see commented on android/autobuild/autobuild_apk_3_copy_files.sh:
```diff
@@ -0,0 +1,44 @@
+echo ""
+echo "" 
+echo "move to ${artifact_deploy_filename}"
+mv ${THIS_JAMULUS_PROJECT_PATH}/android-build/build/outputs/apk/debug/android-build-debug.apk ${THIS_JAMULUS_PROJECT_PATH}/deploy/${artifact_deploy_filename}
```
Why is this a debug .apk? Would it also be possible to use a release apk or is this just for building during normal PR runs?

nefarius2001 replied:
good question... I was very happy to have something that worked, and did not question after that... :)

nefarius2001 replied:
I would like to treat that as a separate issue, or part of "multiplatform apk"

ann0see replied:
Ok. Can you please link this in multiplatform apk?

nefarius2001 replied:
done

ann0see commented on mac/autobuild/codeQL/autobuild_mac_3_copy_files.sh:
```diff
@@ -0,0 +1,7 @@
+
+
+
+echo "no action here"
```
Is this file really needed?

nefarius2001 replied:
it is not necessary, but it helps stylewise.
it obviously isn't in terms of execution, but it makes the 3-step-structure visible & consistent.

ann0see commented:
Building on android seems to take a of time. Any ideas on how to speed it up?

ann0see commented:
@nefarius2001 the flatpak seems to fail. But I'm quite sure that it can soon be removed since we have Jamulus on Flathub.

nefarius2001 commented:
> the flatpak seems to fail. But I'm quite sure that it can soon be removed since we have Jamulus on Flathub.

the error message hints to the recent submodule-thing => fatal: '/__w/google/oboe.git' does not appear to be a git repository
I will deactivate it with a "if False" for now. It can then be removed, when Jamulus is on Flatpak?

> Building on android seems to take a of time. Any ideas on how to speed it up?
I have the suspicion that codeql makes the build slower, but not tested. Apart from that - no. Especially since all preparations are already separated in step 1.

nefarius2001 commented:
!!!!!!!!!!!!!!!!!!!!!!!!!
I enabled the autobuild for master-branch pushes, since codeql gives a warning for pull-requests otherwise.
Giving some !!! here, since it was discussed otherwise and should be discussed in review, if unwanted. I suggest to enable it this way

pljones commented:
> 
> 
> !!!!!!!!!!!!!!!!!!!!!!!!!
> I enabled the autobuild for master-branch pushes, since codeql gives a warning for pull-requests otherwise.
> Giving some !!! here, since it was discussed otherwise and should be discussed in review, if unwanted. I suggest to enable it this way

I think doing what CodeQL is asking for is correct, myself.  A master branch push _can_ happen without a PR.  However, that shouldn't mean it gets away without running CodeQL.  In fact, the merge to master / push to master should be what sets the badge.

ann0see reviewed (commented):
Just came across this issue too. I'm still looking through your code, but in general I think it should be ok. I'm not quite sure, we can remove the flatpak.

@joao-p-marques since we now have Jamulus on flathub, I think we can remove it from the automated build process. Most users would probably rather install it from flathub than download it from GitHub? Do you agree? Of course, the json/yaml file should stay on this repo. Only the automated build would be pointless.

joao-p-marques commented:
@ann0see yes, the flatpak build ends up being duplicated in the 2 repos. Leaving it here as part of the CI process would probably help spot a possible change that makes the Flatpak build fail, and correctly more easily on each individual PR, for example. However, the idea of the flathub repo is to simply build from the latest git release, and I will do some testing each time a new version is released, so it wouldn't be a problem. 
If you prefer, you could even remove the flatpak json manifest, as it is basically duplicated (as yaml) in the flathub repo.

ann0see commented:
Thanks for the answer! I'd wait for a few other answers on this topic and then we can decide later. I'd at least remove it from the official release process.

pljones commented on mac/autobuild/codeQL/autobuild_mac_1_prepare.sh:
```diff
@@ -0,0 +1,48 @@
+###  PROCEDURE  ###
+###################
+
+#USING GITHUB ACTION HERE
```
This comment seems wrong.

nefarius2001 replied:
true ,fixed

nefarius2001 commented:
About flatpak; I'd suggest to remove it from this repo, since it seems to be "broken" by the submodule-paths anyway....

ann0see commented:
I've pushed a beta release tag to the main repo to test it. Let's wait and see what happens. Could you please remove flatpak? Code Ql doesn't seem to work on this PR but maybe that's because it's not yet included this way on master

ann0see commented:
I just saw that you seperated codeQl and the autobuild process for linux and macOS can't you unify them? I don't think we need to run macOS on different versions of Qt.

ann0see reviewed (changes requested):
The quotes in the bash scripts are still missing. I can do it myself, if you want?

ann0see commented:
After looking through it: It seems as if embedding codeQl makes it run a lot longer (the last run lasted around one hour which is too much). I assume that we might not want to include codeQl and packaging in one file then?

nefarius2001 commented:
Thanks for the quotes!

 By having them in one file, it is at least possible to use the output of a codeql run where applicable.
Disabling codeql would also be easy through a Boolean - where would you not have it? I think for a release, build time of some hours is not critical.

on the other hand, maybe two files is more clean, then. I’ll think about it🙂

ann0see commented:
> for a release, build time of some hours is not critical.

Probably not. But for a PR (like this) it might be. Android APK is still building even though  42 minutes have already passed (which is too long for a PR). I think we should solve this problem.

nefarius2001 commented:
43m :)
maybe moving from debug to release build improves the build time. Lets takle one after the other

ann0see commented:
Maybe... Could you please remove the flatpak build?

pljones commented on .github/workflows/autobuild.yml:
```diff
@@ -1,25 +1,53 @@
       - "r*"
       - "latest"
-    #branches:
+    branches:
```
I think I'd still prefer `*` for the CodeQL.

If it's possible to exclude branches from the publish, then perhaps have branches for that (`dev-*` or whatever) that are automatically rejected for pull requests.

ann0see replied:
Yes. That's a great idea!

nefarius2001 replied:
Do I get the idea correctly?
- branch not starting with "dev": CodeQL + artifacts (push or pull-request)
- branch starting with "dev": push: no action at all
- branch starting with "dev": pull request: Error

pljones replied:
Yes.

pljones commented on .github/workflows/autobuild.yml:
```diff
@@ -65,92 +93,190 @@ jobs:
-             asset_name:            jamulus_${{ needs.create_release.outputs.version_name }}_ubuntu_amd64.deb
-           - config_name:           Linux (headless)
+             building_on_os:        ubuntu-20.04
+             cmd1_prebuild:         "sh linux/autobuild/codeQL/auto_1_prepare.sh ${{ needs.create_release.outputs.x_github_workspace }}"
```
We seem to have a lot of scripts to maintain now.  How much of the code is duplicated between them?  Have they been refactored to make the reply to that as close to "none" as possible?

nefarius2001 replied:
There are not more than before, but less scattered and uniformly splitted in 3 steps (to see what effort goes into building and what into preparation).

My idea was to first set an overall structure in this PR, and then be able to clean stages separately. (e.g. some latforms still rely on additional scripts outside the autobuild folders)

pljones commented on android/autobuild/autobuild_apk_2_build.sh:
```diff
@@ -0,0 +1,44 @@
+echo .
+echo .
+echo .
+/opt/android/android-ndk/prebuilt/linux-x86_64/bin/make
```
Seems inconsistent with line 39.

pljones commented on android/autobuild/autobuild_apk_2_build.sh:
```diff
@@ -0,0 +1,44 @@
+###  PARAMETERS  ###
+####################
+
+if [ ! -z "${1}" ]; then
```
Can't this live somewhere and get sourced?

nefarius2001 replied:
I like the idea.
Only difficulty is a clean way to handle that not all scripts need all parameters.

ann0see replied:
This seems to be done now.

pljones commented on linux/autobuild/codeQL/auto_2_build_qt_project.sh:
```diff
@@ -0,0 +1,41 @@
+###  PARAMETERS  ###
+####################
+
+if [ ! -z "${1}" ]; then
```
Not only is there code repeated for the same platform, it's repeated across platforms.

It does argue for restructuring the autobuild to be the top level with platform-specific code in subdirectories.  That would encourage the common code to be shared and only drop in the specifics where needed.

nefarius2001 replied:
I'm still in favor of that.

nefarius2001 replied:
sneak preview:
https://github.com/nefarius2001/jamulus/tree/autobuild_head
https://github.com/nefarius2001/jamulus/commit/784997c747c50b14dd5288b0219e9f57d2e3fc27

will merge to PR after successfull tests

nefarius2001 commented:
BTW, the initial cause for splitting artifacts and codeql in some cases was, that for some unkonown reason the build of the installers did not trigger the codeql-inspection.

Since some topics popped up above, let me know your opinion
- should the PR be open until it cleared everything
- would you have that as an intermediate step, then a next bunch of rework as a new PR

ann0see commented:
I think that this PR should stay open and if it works, we can even merge it and then soon try it (= new jamulus release). Compiling on macOS with Qt 5.9.9 is blocking. 

@pljones what do you think?

@nefarius2001 do you think we can work on a new release process as a team?

ann0see reviewed (commented):
Since it seems to work I will remove my changes request.

nefarius2001 commented:
I’m happy to continue in team here.
Let’s have a short call how to work in parallel and set up a checklist?

ann0see commented:
>  Let’s have a short call how to work in parallel and set up a checklist?

Sure. Just I'd have time now and probably this evening. You can just create a Jitsi link

nefarius2001 commented:
https://meet.jit.si/jamulusbuild

nefarius2001 commented:
follow-up  18:30, same jitsi

nefarius2001 commented:
implemented separate folder structure for autobuild for removing code douplication.
some issues still to remove the code duplications

ann0see commented:
@jamulussoftware/maindevelopers I think it would be great if we could find a time where the release process is discussed with @nefarius2001

ann0see commented:
@nefarius2001 I'm currently also working on getting the linux build split in headless and gui. I think there are multiple bugs in the linux generation scripts. Nevertheless, I'll have a look at https://github.com/ann0see/jamulus/tree/LinuxUnify

nefarius2001 commented:
- removed the code duplication by calling one script from all others
- moved all scripts in separate folder structure "autobuild"

First, I need to figure out why the same action passes for my repo, but fails on PR...
https://github.com/jamulussoftware/jamulus/pull/978/checks?check_run_id=1922134046
https://github.com/nefarius2001/jamulus/runs/1922133793?check_suite_focus=true
Any ideas?

Second, I'd ask you to merge this to master as a basis (after fixed)
The other open improvements can then be tackled in independant, smaller issues/PRs
Let me know if there are other issues with this PR

ann0see commented:
@hoffie is already working on a fix for the headless build.

ann0see commented on .github/workflows/autobuild.yml:
```diff
@@ -65,100 +88,140 @@ jobs:
+           - config_name:           AndroidAPK (artifact+codeQL)
+             target_os:             android
+             building_on_os:        ubuntu-20.04
+             cmd1_prebuild:         "sh ./autobuild/android/autobuild_apk_1_prepare.sh "
```
Question: Why did you add a space after the quote here? (applies to multiple other commands) below.

ann0see replied:
Done

nefarius2001 replied:
you mean the offset of the last quote?
Historical, the jamulus_project_path used to be passed as a "normal" parameter to the call, before I decided that a named environment variable will be more intuitive to read for others. Also more flexible if additional parameters are needed in some distant future.

ann0see replied:
> you mean the offset of the last quote?

Yes. 

I don't see any benefit from adding a space here... I'd remove it, if possible. You can always add a space here.

ann0see replied:
Ok. Thanks

ann0see commented on .github/workflows/autobuild.yml:
```diff
@@ -65,100 +88,140 @@ jobs:
-           - config_name:           MacOS
+             building_on_os:        ubuntu-20.04
+             cmd1_prebuild:         "sh ./autobuild/linux/autobuild_deb_1_prepare.sh"
+             cmd2_build:            "./autobuild/linux/autobuild_deb_2_build.sh"
```
Why no "sh" here?

ann0see replied:
Fixed.

nefarius2001 replied:
the scripts 2&3 need "bash" for some of the commands.
invoking 1 with bash leads to some permission-errors

using sh for step one and bash for the other two works well.

nefarius2001 replied:
added a comment

ann0see replied:
Ok. Great.

ann0see commented on autobuild/android/autobuild_apk_1_prepare.sh:
```diff
@@ -0,0 +1,99 @@
+export DEBIAN_FRONTEND="noninteractive"
+echo "::set-env name=DEBIAN_FRONTEND::${DEBIAN_FRONTEND}"
+
+sudo apt-get  update
```
Todo: remove space after apt-get

ann0see replied:
Ok. Seems to be the style of this file.

ann0see commented on autobuild/android/install-qt.sh:
```diff
@@ -0,0 +1,355 @@
+#!/usr/bin/env bash
```
Not blocking, but maybe one time we could move over to aqt.

nefarius2001 replied:
aqt for all, but later, one after the overall structure

ann0see replied:
Ok. Let's resolve this and work on it after this got merged.

ann0see commented on autobuild/linux/autobuild_deb_2_build.sh:
```diff
@@ -0,0 +1,18 @@
+
+cd "${THIS_JAMULUS_PROJECT_PATH}"/distributions
+
+sh ./build-debian-package-auto.sh
```
Note to self: todo; not blocking

ann0see commented on autobuild/mac/artifacts/autobuild_mac_2_build.sh:
```diff
@@ -0,0 +1,19 @@
+cd "${THIS_JAMULUS_PROJECT_PATH}"
+
+echo "Run deploy script..."
+sh ${THIS_JAMULUS_PROJECT_PATH}/mac/deploy_mac.sh
```
Todo: quoting

ann0see replied:
Done.

ann0see commented on autobuild/mac/codeQL/autobuild_mac_1_prepare.sh:
```diff
@@ -0,0 +1,11 @@
+###################
+
+echo "Install dependencies..."
+brew install qt5
```
Todo: use aqt and the same process as in the release process.

pljones replied:
There should only be one "Install Qt" (per platform).  If it's always needed, always run it (as a workflow step).  One workflow.  Some steps always run (i.e. those that prepare for what happens later).  Some steps only run under certain conditions (branch name, action, target - CodeQL only runs sometimes, publish only runs sometimes).

("target" - this doesn't just mean platform, it includes combinations like "headless server on debian x64".)

ann0see commented on autobuild/windows/autobuild_windowsinstaller_1_prepare.ps1:
```diff
@@ -1,7 +1,11 @@
-# Sets up the environment for autobuild on Windows
+# Powershell
```
Are you sure this comment is better than the one from before?

nefarius2001 replied:
line 3 is "# autobuild_1_prepare: set up environment, install Qt & dependencies" << thats the one to compare, and I think it is more concrete.

line 1 mimmics the shebang of other files, so mac/linux/windows scripts have the same header format

ann0see replied:
Hmm. Ok. Will resolve this too.

ann0see commented on distributions/io.jamulus.Jamulus.prepare.py:
```diff
@@ -1,5 +1,6 @@
 #!/usr/bin/python3
```
Probably you could revert all the changes in this file since we don't use flatpak anymore.

ann0see replied:
Removed this file

ann0see reviewed (changes requested):
Almost finished. Will look into changing a few things here (and I'll update the branch to the current master branch because of the broken headless build). Afterwards I'll do a new beta release.

ann0see reviewed (approved):
Aside from the slow CodeQl build, I'd say this is ready once the comments above are answered. There are still some open points, but they're not part of this PR. Therefore I'll approve this. If the CI doesn't fail I'd be ready to merge it.

pljones commented on autobuild/ensure_THIS_JAMULUS_PROJECT_PATH.sh:
```diff
@@ -0,0 +1,33 @@
+  #python -c "import os; print(os.path.abspath(\"${THIS_JAMULUS_PROJECT_PATH}\"));"
+  echo "THIS_JAMULUS_PROJECT_PATH exists (absolute): ${THIS_JAMULUS_PROJECT_PATH}"
+else
+    echo "ERROR: non-existing THIS_JAMULUS_PROJECT_PATH: ${THIS_JAMULUS_PROJECT_PATH}"
```
Non-blocking: `-d` checks explicitly for a directory, so the failure message should make that clear.  Something like:
```
echo "ERROR: ${THIS_JAMULUS_PROJECT_PATH} must reference a directory."
```

nefarius2001 replied:
done

pljones commented on autobuild/linux/autobuild_deb_1_prepare.sh:
```diff
@@ -0,0 +1,17 @@
+
+echo "Update system..."
+sudo apt-get -qq update
+# We don't upgrade the packages. If this is needed, just uncomment this line
```
No, just remove these two commented lines.  We need apt updated and then the install should pull the required packages correctly.

nefarius2001 replied:
can do, no strong own opinion on this

pljones commented on autobuild/mac/artifacts/autobuild_mac_1_prepare.sh:
```diff
@@ -0,0 +1,20 @@
+###################
+
+echo "Install dependencies..."
+#brew install qt5
```
(non-blocking) lose the comments - they'll be in git history should we ever want them.

nefarius2001 replied:
done

pljones commented on autobuild/mac/codeQL/autobuild_mac_2_build.sh:
```diff
@@ -0,0 +1,31 @@
+
+
+echo "Building... qmake"
+if [-x /Users/runner/work/jamulus/jamulus/Qt/5.15.2/clang_64/bin/qmake]
```
Did you try it without specifying the full path like this?  I found it worked just doing `qmake`, after running the install in the CodeQL workflow.  If it does, it would make it less painful moving from 5.15.2 to 5.9 for the MacOS build.

pljones commented on autobuild/mac/codeQL/autobuild_mac_3_copy_files.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
```
See, this is where the current workflow design seems broken.

It should know this is a CodeQL run and not schedule a copy step.

nefarius2001 replied:
I removed the file, since that is not my focus at the moment.

pljones commented:
I think there's more to do - but it can wait till after this gets merged.  Continuous Improvement means taking "good enough" and making it better, later.  This is definitely good enough and better that we have.

pljones commented:
#964 though.  Is that covered?

ann0see commented:
Yes. This is included as far as I know.

ann0see commented:
@pljones I'd wait a bit until @nefarius2001 is finished and afterwards, you can merge it.

pljones commented:
> 
> 
> Yes. This is included as far as I know.

Ah - the confusion is around the CodeQL vs main build - one uses 5.12.2, the other 5.9.9.  That can wait (we've been using 5.12.2 for CodeQL up until now) but needs fixing, as we need to verify against the same library base as the build.

ann0see commented:
Ok. Will merge this now.

ann0see commented:
There are still some todo items, but I think it's ready. Next big step:

Start the release process!
